projectq/ops/_basics.py

```diff
@@ -101,6 +101,8 @@ def __init__(self):
                 self.set_interchangeable_qubit_indices([[0,1],[2,3,4]])
         """
         self.interchangeable_qubit_indices = []
+        self._commutable_gates = []
+        self._commutable_circuit_list = []
```

Comment:
I don’t think this would work because we want to be able to choose different values for commutable_circuit_list and commutable_gates for gates wrapped in a ControlledGate. 
For instance I have defined a get_commutable_circuit_list function in the XGate class which takes the parameter n as the number of control qubits and can return different commutable circuit lists depending on n. As far as I can tell the only way the optimizer can access the number of controls is at run time, by evaluating len(self.control_qubits) (see _command.py::Command::is_commutable). Therefore the _commutable_gates and _commutable_circuit_list attributes need to be instance attributes.

Reply:
Ok, makes sense.